### PR TITLE
[generate] deprecate forced ids processor

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

First stage of deprecation for the `ForceTokensLogitsProcessor`. This logits processor was exclusively used for Whisper. However, the PR #29485 removed the last calls to this processor from Whisper, in favour of `decoder_input_ids` set internally by generate, meaning this processor can now be removed.

### CODE REVIEW

**Review by ArthurZucker — APPROVED** [review]
Glad that we removed it!

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -1745,8 +1746,14 @@ class ForceTokensLogitsProcessor(LogitsProcessor):
     ```
     """
 
-    def __init__(self, force_token_map: List[List[int]]):
+    def __init__(self, force_token_map: List[List[int]], _has_warned: Optional[bool] = False):
         self.force_token_map = dict(force_token_map)
+        if not _has_warned:
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/generation/logits_process.py:1751]: This argument is to distinguish between when the user has passed `forced_ids` to generate, in which case we throw a warning explaining so:
https://github.com/huggingface/transformers/blob/4a604ed2cc6ed788f5946de20f497cb2dac10d7f/src/transformers/generation/utils.py#L921-L926

OR if the user instantiates the forced logits processor standalone. Again, we give a descriptive warning explaining this:
https://github.com/huggingface/transformers/blob/4a604ed2cc6ed788f5946de20f497cb2dac10d7f/src/transformers/generation/logits_process.py#L1751-L1756

=> we don't want to trigger the second warnings if the user does option 1, since the processor has been instantiated automatically. Thus, we have a separate warning for each case.

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -1745,8 +1746,14 @@ class ForceTokensLogitsProcessor(LogitsProcessor):
...
+            # TODO(Sanchit): remove this processor entirely in v4.40
+            warnings.warn(
+                "This `ForceTokensLogitsProcessor` has been deprecated and will be removed in v4.40. Please pass `input_ids` or `decoder_input_ids` to the generate method directly.",
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/generation/logits_process.py]: Resolved in https://github.com/huggingface/transformers/pull/29487/commits/976f07aea82e23a339273c613f09e1029325037a
